### ISSUE TITLE
Request rebinding clears default route parameters

### ISSUE THREAD

**Issue by stidges** [laravel/framework#30291]
- Laravel Version: 6.x
- PHP Version: 7.3.x
- Database Driver & Version: N/A

### Description:

Whenever the request is rebound in the container, any default route parameters defined will be lost, because the [RouteUrlGenerator](https://github.com/laravel/framework/blob/80914c430fb5e49f492812d704ba6aeec03d80a2/src/Illuminate/Routing/UrlGenerator.php#L688) is cleared. This is mostly occurring during testing, as any HTTP call will rebind the request in the container.

This could possible be fixed by temporarily storing the defined default parameters in a variable, clearing the `$routeGenerator` property, and then reassigning the default parameters. If this sounds like a good solution I'd be happy to open a PR!

### Steps To Reproduce:

The following test demonstrates the issue.

```php
<?php

namespace Tests\Feature;

use Illuminate\Support\Facades\Route;
use Illuminate\Support\Facades\URL;
use Tests\TestCase;

class ExampleTest extends TestCase
{
    public function testRouteParameterDefaults()
    {
        Route::get('/bar', function () {})->prefix('{prefix}')->name('test');
        Route::getRoutes()->refreshNameLookups();
        URL::defaults(['prefix' => 'foo']);
        $url = route('test');
        // -> "http://localhost/foo/bar"

        $response = $this->get($url);

        $this->assertEquals($url, route('test'));
        // -> Illuminate\Routing\Exceptions\UrlGenerationException
        //    : Missing required parameters for [Route: test] [URI: {prefix}/bar].
    }
}
```

**Comment by driesvints** [comment]
I'm not really sure if the above test makes sense? Why would you resolve the route again? Can you maybe post a more real-life example?

**Comment by stidges** [comment]
Sure! In my application it's actually a route group with a dynamic domain, and I want to assert that I'm redirected to another page within that group, e.g.:

```php
Route::domain('app.{domain}')->group(function () {
    Route::post('/', function () {
        // Validate and login the user
        return redirect()->route('dashboard');
    })->name('login');

    Route::get('/dashboard', function () {})->name('dashboard');
});
Route::getRoutes()->refreshNameLookups();

URL::defaults(['domain' => 'mydomain.test']);

$response = $this->post(route('login'), [
    'username' => 'test-user',
    'password' => 'password',
]);

$response->assertRedirect(route('dashboard'));
```

**Comment by driesvints** [comment]
What is `Route::getRoutes()->refreshNameLookups();` doing there?

**Comment by stidges** [comment]
I've added that in to make the testcase work on its own, otherwise you'd get a `Route [login] not defined` exception because the routes are defined after the framework has booted. In my application the routes are actually defined in a routes file, in which case Laravel's [RouteServiceProvider would handle that automatically](https://github.com/laravel/framework/blob/f61b787d88505d94d4ca691d43307518aed3dbd9/src/Illuminate/Foundation/Support/Providers/RouteServiceProvider.php#L39)

**Comment by driesvints** [comment]
Ah I see now. Unfortunately we don't support defining routes in the test case itself. Does it also not work when you define them in a routes file?

**Comment by stidges** [comment]
It gives the same exception after using `route` helper after the request was made, the initial call  works as expected. This is because the `call` method [creates and sends a new request](https://github.com/laravel/framework/blob/f61b787d88505d94d4ca691d43307518aed3dbd9/src/Illuminate/Foundation/Testing/Concerns/MakesHttpRequests.php#L433) through the HTTP kernel, which [rebinds the request](https://github.com/laravel/framework/blob/f61b787d88505d94d4ca691d43307518aed3dbd9/src/Illuminate/Foundation/Http/Kernel.php#L142). So after any test request the route parameter defaults will no longer be available.

**Comment by driesvints** [comment]
I just tested this out and it indeed also gives me the route undefined error. But tbh this is expected behavior. I don't see a reason why you would want to recreate the route again after you've run your test. You usually only run assertions after executing your test and do all arrangements before you execute it.

So I don't consider this a bug, sorry. You're always free to open up an issue on the ideas repo to see if anyone else is interested in an improvement or send in a PR to see if it gets accepted. Thanks.

**Comment by stidges** [comment]
I am not actually recreating the route after assertions, I'm only testing if the user is redirected to the expected route which is a common test case. The routes I'm using in my test are already defined in a routes file, the only thing that goes wrong is the default route parameters being lost in the process. 

I've already implemented a workaround in my application, but thought it could be something that we may want to address at the framework level. I understand that it is a bit of a specific problem, so I'm fine with implementing the workaround on the application level. For completeness, in case it may help someone else, here's (a simplified version of) what I came up with:

```php
// app/Providers/RouteServiceProvider.php
public function boot()
{
    $this->app['url']->defaults(['domain' => 'mydomain.test']);

    $this->app->rebinding('request', function ($app, $request) {
        // Redefine default route parameters when request is rebound.
        $app['url']->defaults(['domain' => 'mydomain.test']);
    });

    parent::boot();
}
```
